src/Microsoft.Diagnostics.DebugServices.Implementation/ModuleService.cs

```diff
@@ -0,0 +1,572 @@
+        /// <param name="version">the PE version or null</param>
+        /// <param name="flags">module flags</param>
+        /// <returns>PEImage instance or null</returns>
+        private PEImage GetPEInfo(bool isVirtual, ulong address, ulong size, ref PdbInfo pdbInfo, ref VersionInfo? version, ref Module.Flags flags)
+        {
+            Stream stream = MemoryService.CreateMemoryStream(address, size);
+            try
+            {
+                stream.Position = 0;
+                var peImage = new PEImage(stream, leaveOpen: false, isVirtual);
+                if (peImage.IsValid)
+                {
+                    flags |= Module.Flags.IsPEImage;
+                    flags |= peImage.IsManaged ? Module.Flags.IsManaged : Module.Flags.None;
+                    pdbInfo = peImage.DefaultPdb;
+                    if (!version.HasValue)
+                    {
+                        FileVersionInfo fileVersionInfo = peImage.GetFileVersionInfo();
+                        if (fileVersionInfo != null)
+                        {
+                            version = fileVersionInfo.VersionInfo;
+                        }
+                    }
+                    flags &= ~(Module.Flags.IsLoadedLayout | Module.Flags.IsFileLayout);
+                    flags |= isVirtual ? Module.Flags.IsLoadedLayout : Module.Flags.IsFileLayout;
+                    return peImage;
+                }
+            }
+            catch (Exception ex) when (ex is BadImageFormatException || ex is EndOfStreamException || ex is IOException)
+            {
+                Trace.TraceError($"GetPEInfo: loaded {address:X16} exception {ex.Message}");
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Finds or downloads the module and creates a PEReader for it.
+        /// </summary>
+        /// <param name="module">module instance</param>
+        /// <returns>reader or null</returns>
+        internal PEReader GetPEReader(IModule module)
+        {
+            string downloadFilePath = null;
+            PEReader reader = null;
+
+            if (File.Exists(module.FileName))
+            {
+                downloadFilePath = module.FileName;
+            }
+            else 
+            { 
+                if (SymbolService.IsSymbolStoreEnabled)
+                {
+                    SymbolStoreKey key = PEFileKeyGenerator.GetKey(Path.GetFileName(module.FileName), (uint)module.IndexTimeStamp, (uint)module.IndexFileSize);
+                    if (key != null)
+                    {
+                        // Now download the module from the symbol server
+                        downloadFilePath = SymbolService.DownloadFile(key);
+                    }
+                }
+            }
+
+            if (!string.IsNullOrEmpty(downloadFilePath))
+            {
+                Trace.TraceInformation("GetPEReader: downloading {0}", downloadFilePath);
```

Comment:
At this point the file is fully downloaded (`DownloadFile` is a blocking call). This trace is not accurate.